### PR TITLE
Remove the usage of `SELFDESTRUCT` in `BytecodeStorage`

### LINKED ISSUE #451: Remove usage of SELFDESTRUCT in bytecodestorage library
I don't think that this is urgent AFAICT, but filing now that https://eips.ethereum.org/EIPS/eip-6049 is out in the wild.

Additional context: 

- https://twitter.com/TimBeiko/status/1616112768272457729?s=20&t=XXcwOvFZ969YwSzKEyagkQ
- https://github.com/ethereum/execution-specs/pull/689

cc @ryley-o

### PR DESCRIPTION
**Pull request by jakerockland**
## Description of the change

Purge the use of purging! 

Closes #451 

~Only merge after #665 – also, do not merge until review feels coherent between this and the TBD PR that addresses https://github.com/ArtBlocks/artblocks-contracts/issues/660~

Do not merge until this changeset looks good _in addition to_ https://github.com/ArtBlocks/artblocks-contracts/pull/670, as both should be merged at the same time.

---

Note that the approach of maintaining multiple versioned files of our library is taken in order to allow for testing backwards compatible reads, though backwards compatible reads cannot be reliably added until addressing https://github.com/ArtBlocks/artblocks-contracts/issues/660 which is being taken on in a companion PR to this one. These PRs should be reviewed in tandem.

### CODE REVIEW

File: contracts/libs/0.8.x/BytecodeStorageV1.sol [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    uint256 internal constant ADDRESS_OFFSET = 1;
+
+    /*//////////////////////////////////////////////////////////////
+                           WRITE LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Write a string to contract bytecode
+     * @param _data string to be written to contract. No input validation is performed on this parameter.
+     * @return address_ address of deployed contract with bytecode containing concat(deployer-address, data)
```
Comment (ryley-o) [inline contracts/libs/0.8.x/BytecodeStorageV1.sol:49]:
```suggestion
     * @return address_ address of deployed contract with bytecode containing concat(invalid opcode, deployer-address, data)
```

File: contracts/libs/0.8.x/BytecodeStorageV1.sol [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        require(address_ != address(0), "ContractAsStorage: Write Error");
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               READ LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Read a string from contract bytecode
+     * @param _address address of deployed contract with bytecode containing concat(deployer-address, data)
```
Comment (ryley-o) [inline contracts/libs/0.8.x/BytecodeStorageV1.sol:110]:
```suggestion
     * @param _address address of deployed contract with bytecode containing concat(invalid opcode, deployer-address, data)
```

File: contracts/libs/0.8.x/BytecodeStorageV1.sol [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            // store length of data in first 32 bytes
+            mstore(data, size)
+            // copy code to memory, excluding the deployer-address
+            extcodecopy(_address, add(data, 0x20), DATA_OFFSET, size)
+        }
+    }
+
+    /**
+     * @notice Get address for deployer for given contract bytecode
+     * @param _address address of deployed contract with bytecode containing concat(deployer-address, data)
```
Comment (ryley-o) [inline contracts/libs/0.8.x/BytecodeStorageV1.sol:150]:
```suggestion
     * @param _address address of deployed contract with bytecode containing concat(invalid opcode,  deployer-address, data)
```

**Review by ryley-o — APPROVED** [review]
LGTM, with agreement of all the caveats listed in this PR's description (i.e. let's not deploy any Engine contracts until #660 is also included in this V1 BytecodeStorage library 🙏)

**Comment by jakerockland** [comment]
> LGTM, with agreement of all the caveats listed in this PR's description (i.e. let's not deploy any Engine contracts until #660 is also included in this V1 BytecodeStorage library 🙏)

Very much agreed – super important not to deploy any contracts using this modified structure until #660 is resolved (will have an initial draft PR up for that shortly here, after which we can look at these two changesets in tandem).

**Review by yoshiwarab — APPROVED** [review]

**Pull request merged**